Lib/socket.py

```diff
@@ -127,7 +127,33 @@ def _intenum_converter(value, enum_klass):
     __all__.append("errorTab")
 
 
-class _GiveupOnSendfile(Exception): pass
+def _check_sendfile_params(sock, file, offset, count):
+    if 'b' not in getattr(file, 'mode', 'b'):
+        raise ValueError("file should be opened in binary mode")
+    if not sock.type == SOCK_STREAM:
+        raise ValueError("only SOCK_STREAM type sockets are supported")
+    if count is not None:
+        if not isinstance(count, int):
+            raise TypeError(
+                "count must be a positive integer (got {!r})".format(count))
+        if count <= 0:
+            raise ValueError(
+                "count must be a positive integer (got {!r})".format(count))
+
+
+def _prepare_sendfile(file, count):
+    try:
+        fileno = file.fileno()
```

Comment:
We must support passing an `int` fd to `loop.sock_sendfile`, so that it behaves similar to `loop.sock_sendall` etc. Look at how `selectors.py` calls fileno, and replicate it here.

Reply:
`file` here is sent file object, a second parameter.
Supporting int file descriptors for sent files makes sense, but we should do it for `socket.sendfile` and `loop.sock_sendfile` in parallel.
Let's implement it in the next issue/PR.

Reply:
I'm fine with the `file` arg.  Let's add that in a separate PR.

I'm more concerned with the `sock` parameter now. In `loop._sock_sendfile` helper method, you do `fd = sock.fileno()`, which means that it's not possible to call `loop.sock_sendfile(sock_FD, file)`.

And since it's currently possible to call `loop.sock_sendall(sock_FD, ...)`, we should ensure that the new `loop.sock_sendfile()` API is designed similarly.

Reply:
fixed